lxd: Mark images as cached consistently

nmezhenskyi opened the pull request:
This PR makes marking images as cached consistent among projects that have `features.images` disabled.

Fixes https://github.com/canonical/lxd/issues/16576.

## Changes:
- If `features.images` is disabled for the project, associate the image with the `default` project inside of the `ensureDownloadedImageFitWithinBudget()` function. This ensures that `cached=yes` is set for a new image downloaded implicitly using `lxc init <image> <instance_name>`.
-  If `features.images` is disabled for the project, associate the image with the `default` project during `lxc copy`. This ensures that `cached=no` is set for a new image downloaded explicitly using `lxc image copy`.
- Correctly associate profile IDs for images that are shared between the `default` project and custom projects with `features.images=false`.
- Enforce consistent use of `shared.IsFalseOrEmpty()` when checking `features.images` value for a project.

nmezhenskyi committed:
lxd/instances_post: If project features.images is disabled, associate image with default project

Signed-off-by: Nikita Mezhenskyi <nikita.mezhenskyi@canonical.com>

tomponline commented:
@nmezhenskyi please can you add a test for this, ta

nmezhenskyi committed:
lxd/images: Correctly associate images and profiles with projects

Signed-off-by: Nikita Mezhenskyi <nikita.mezhenskyi@canonical.com>

nmezhenskyi commented on lxd/db/images.go:
```diff
@@ -644,7 +644,7 @@ func (c *ClusterTx) UpdateImage(ctx context.Context, id int, fname string, sz in
 		}
 
-		sql = `INSERT INTO images_profiles (image_id, profile_id) VALUES (?, ?)`
+		sql = `INSERT OR IGNORE INTO images_profiles (image_id, profile_id) VALUES (?, ?)`
```
This is needed because the profile could already be associated with the image by previously calling `CreateImage()` function.   

`CreateImage()` itself also handles this query using `INSERT OR IGNORE`:
https://github.com/canonical/lxd/blob/5e69be9583bf52f7803a2cc8f5ba8a26dcd2beda/lxd/db/images.go#L751-L768

nmezhenskyi replied:
> arent the images_profiles rows deleted just above this?

They are, but only for a single project associated with the image. There is an edge case where this does not work: when a project has `features.images=false`, but `features.profiles=true`. This means that the image is associated to both the `default` project's profiles and to the other custom project's profiles. The `DELETE` query will only delete the association with the custom project's profiles.

Should I fix this query as well?

tomponline replied:
> Should I fix this query as well?

What do you think is best, leaving it as insert ignore?
should we do that instead of the other deletion?

nmezhenskyi replied:
> What do you think is best, leaving it as insert ignore?

Hm, I'm still trying to understand the correct approach for updating image profiles in `UpdateImage()`.

Using `insert ignore` seems to work in this particular case, but I'm afraid it can introduce a subtle bug when there are multiple projects with `features.profiles=true` and `features.images=false`. 

Let's say we have three projects: `default`, `project1`, `project2`.

`project1` and `project2` have `features.profiles=true` and `features.images=false`. This means they have their own profiles, but they share images with the `default` project.

All three projects have a single profile named `profile1`. The `default` project has image named `image1` and it is associated with `profile1`.

 This means that the image actually has profile associations like these in the database:

**images_profiles rows**:

| Image | Profile |
| -------- | ------- |
| default_project_image1 | default_project_profile1 |
| default_project_image1 | project1_profile1 |
| default_project_image1 | project2_profile1 |

Now, the problem occurs when we want to remove `profile1` from `image1`, the `delete` query in the `UpdateImage()` will only do it for the `default` project.

tomponline replied:
Right makes sense, and then we'd expect the caller to figure out all the profile IDs across all projects that reference this image.

nmezhenskyi replied:
I've updated the approach as described, but it seems to have broken other tests. I'm looking into this.

nmezhenskyi committed:
test: Check that images are being cached consistently between projects

Signed-off-by: Nikita Mezhenskyi <nikita.mezhenskyi@canonical.com>

tomponline commented on test/suites/image.sh:
```diff
@@ -398,3 +398,76 @@ run_images_public() {
+  lxc delete c1
+
+  echo "==> Check that the locally saved image is marked as cached."
+  lxc image info "${fingerprint}" | grep -xF "Cached: yes"
```
Can you use `lxc image get-property` here?

tomponline commented on test/suites/image.sh:
```diff
@@ -398,3 +398,76 @@ run_images_public() {
+  lxc init images:alpine/edge c1
+
+  local fingerprint
+  fingerprint="$(lxc config show c1 | sed -En 's/^.*volatile.base_image: (.*)$/\1/p')"
```
can you use `lxc config get` here?

nmezhenskyi committed:
lxd/db/images: Rework UpdateImage function to completely overwrite profile associations for the image

This is needed to ensure that profile associations from multiple
projects are correctly removed when updating image profiles.

Signed-off-by: Nikita Mezhenskyi <nikita.mezhenskyi@canonical.com>

nmezhenskyi committed:
lxd/images: Add resolveProfileIDs function

resolveProfileIDs finds profile IDs in other projects matching the given names.
It searches the "default" project and all projects where "features.images" is false.

Signed-off-by: Nikita Mezhenskyi <nikita.mezhenskyi@canonical.com>

nmezhenskyi committed:
lxd/images: Update calls to UpdateImage function to match new signature

Signed-off-by: Nikita Mezhenskyi <nikita.mezhenskyi@canonical.com>

simondeziel commented on test/suites/image.sh:
```diff
@@ -398,3 +398,76 @@ run_images_public() {
+  echo "==> Test that images are being cached consistently between projects."
+
+  echo "==> Create a new instance in the default project with a remote image that is not stored locally."
+  lxc init images:alpine/edge c1
```
You mean a dedicated LXD image server on which we'd host our `testimage` and a copy of the `ubuntu-minimal-daily:24.04`?

Not that I know but it sounds like an interesting idea.

tomponline replied:
@nmezhenskyi take a look at `test_image_auto_update` and how we used the default lxd server as an image remote to auto update the built-in test image, rather than depending on an external image server.

simondeziel reviewed (dismissed):
LGTM and even with the repeated image fetching, it still runs inside ~20s on `ceph` so not bad at all.

nmezhenskyi commented on lxd/images.go:
```diff
@@ -5086,3 +5086,32 @@ func createTokenResponse(s *state.State, r *http.Request, projectName string, fi
+	profileIDs := make([]int64, 0)
+
+	for _, projectName := range projects {
+		if projectName == currentProject {
```
After looking at it again, I see that there is still a use case for separating fetching profile IDs for the current project and other projects.

When we fetch profile IDs for the current project before calling `getProfileIDsFromOtherProjects()`, we also ensure that all requested profiles actually exist within the current project. Otherwise, we return an error specifying that a profile is missing.

However, this is not a requirement for fetching profile IDs from other projects because other projects are allowed to not have the same profiles.

So the reason we exclude `currentProject` inside of `getProfileIDsFromOtherProjects()` function is to avoid having duplicate IDs in the result list.

tomponline commented on test/suites/image.sh:
```diff
@@ -398,3 +398,93 @@ run_images_public() {
+  lxc project create --storage "${defaultPoolName}" p1
+
+  echo "==> Switch to project p1."
+  lxc project switch p1
```
For tests I recommend using `--project` flag rather than persistent switching as its clearer when reviewing the tests which project each action is occurring in rather than having to remember the state.

nmezhenskyi commented on test/suites/image.sh:
> Can we test (if not already) whether the same image can be cached in one project but not cached in another project (when features.images=true)?

@tomponline Can you please expand on what is the expected behavior here?

Currently I see that if we do `lxc init images:alpine/edge c1` in the `default` project and then do the same for project `p1` with `features.images=true`, both projects will show that their image is cached (`lxc image info <image>` and `lxc image info <image> --project p1`). Is that wrong?

tomponline replied:
That is correct. Both images were created in the project as a side effect of doing `lxc init`.

My question was for a different scenario:

1. lxc init images:alpine/edge c1 in the default project - expect cached image in default project.
2. lxc image copy images:alpine/edge local: in the p1 project - expect non-cached image in p1 project (and expect that same image in default project is still marked as cached).

nmezhenskyi replied:
> My question was for a different scenario

Now I understand, thanks. I've added a check for this now.

nmezhenskyi replied:
> Also, related to the Cached property is the expires_at property, can you expand the tests to check that is correctly being set too?

@tomponline While working on expanding the tests to check for the `expires_at`, I've noticed that we never set this property for images downloaded from `https://images.lxd.canonical.com`, but we do set it for images downloaded from `https://cloud-images.ubuntu.com/minimal/daily`. Is this a bug?

nmezhenskyi replied:
> which part of the code is contributing to the difference?

@tomponline After today's meeting I've confirmed that `expires_at` field is set by the remote server itself.

Here's the relevant section of code inside of the `ImageDownload` function:
 https://github.com/canonical/lxd/blob/936bd063af43f710f2c21ba8c9d26299a3b0b0ab/lxd/daemon_images.go#L113-L129
 
`info, _, err = remote.GetImage(fp)` returns the image metadata which includes the `expires_at` field. Then, we just use this value for the downloaded image.

tomponline commented on lxd/images.go:
```diff
@@ -612,9 +612,17 @@ func imgPostRemoteInfo(ctx context.Context, s *state.State, req api.ImagesPost,
 		}
 
+		// Find IDs of profiles that have matching names in other projects.
+		otherProfileIDs, err := getProfileIDsFromOtherProjects(ctx, tx, profileProject, req.Profiles)
```
Also please can you check, and add if not, tests for checking this profile association behaviour. Thanks

nmezhenskyi replied:
> Right makes sense, and then we'd expect the caller to figure out all the profile IDs across all projects that reference this image [that match the requested profile names].

Initially I approached this by only fetching profiles of projects that have `features.images=false` and only when determined that the `current` project has `features.images=false` or if that's the `default` project. 

However, this approach broke some existing tests that copy local images between projects. What's interesting, it seems to break only when copying a locally created image like Busybox and works correctly with images downloaded remotely. I suspect there might be another bug. I'm investigating this.

nmezhenskyi replied:
> There's an edge case when copying an image from `default` project into another project with `features.images=false`. Logically, this should be a no-op because the image still remains in the `default` project, but current implementation doesn't handle this.

Or should it be an error? @tomponline

nmezhenskyi committed:
lxd/project/limits/permissions: Fix GetImageSpaceBudget to correctly get features.images value

Signed-off-by: Nikita Mezhenskyi <nikita.mezhenskyi@canonical.com>

nmezhenskyi committed:
lxd/api_project: Fix getting features.images value

Correctly evaluate the "features.images" value when creating and
updating a project.

Signed-off-by: Nikita Mezhenskyi <nikita.mezhenskyi@canonical.com>

nmezhenskyi committed:
lxd/db/cluster/projects: Add GetProjectsSharingDefaultImages function

GetProjectsSharingDefaultImages returns the names of all projects using
images stored in the "default" project. That is, it returns all projects
with "features.images=false" and the "default" project itself.

Signed-off-by: Nikita Mezhenskyi <nikita.mezhenskyi@canonical.com>

nmezhenskyi committed:
lxd/db/cluster/projects: Fix comment for GetProjectNames

Signed-off-by: Nikita Mezhenskyi <nikita.mezhenskyi@canonical.com>

nmezhenskyi committed:
lxd/db/cluster/profiles: Add GetProfileIDsByProjectAndName function

GetProfileIDsByProjectAndName finds profile IDs matching specific names within
a specific list of projects.

Signed-off-by: Nikita Mezhenskyi <nikita.mezhenskyi@canonical.com>

tomponline reviewed (approved):
Lovely stuff thanks!
